### PR TITLE
Load history and first set of new messages with 1 API request

### PR DESCRIPTION
**Pull request by nickvergessen**
Ref #8710 

### 🚧 TODO

- [x] Add endpoint to get the message context
  - [ ] Integration tests
- [x] Use for first load of messages when entering a chat
- [x] Revert not setting the read marker
- [ ] JS/Frontend Test
  - [x] Check guests behaviour
  - [x] Check self-joined users behaviour
  - [x] Check revisiting a check room
  - [x] Check behaviour with more than 100 unread messages (context end)
  - [x] Check behaviour with more than 300 unread messages (first future end)

### 🏁 Checklist

- [ ] ⛑️ Tests (unit and/or integration) are included
- [x] 📘 API documentation in `docs/` has been updated or is not required
- [ ] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required
- [x] 🔖 Capability is added or not needed

### CODE REVIEW

File: lib/Controller/ChatController.php [hunk 1]
```diff
@@ -738,7 +764,7 @@ public function clearHistory(): DataResponse {
 	 * @return DataResponse
 	 */
 	public function setReadMarker(int $lastReadMessage): DataResponse {
-		$this->participantService->updateLastReadMessage($this->participant, $lastReadMessage);
+		// FIXME RESTORE $this->participantService->updateLastReadMessage($this->participant, $lastReadMessage);
```
Comment (nickvergessen) [inline lib/Controller/ChatController.php:767]: Temp

File: lib/Controller/ChatController.php [hunk 2]
```diff
@@ -438,16 +438,20 @@ public function receiveMessages(int $lookIntoFuture,
 		$attendee = $this->participant->getAttendee();
 		if ($lookIntoFuture && $setReadMarker === 1 &&
 			$lastKnownMessageId > $attendee->getLastReadMessage()) {
-			$this->participantService->updateLastReadMessage($this->participant, $lastKnownMessageId);
+			// FIXME RESTORE $this->participantService->updateLastReadMessage($this->participant, $lastKnownMessageId);
```
Comment (nickvergessen) [inline lib/Controller/ChatController.php:441]: Temp

**Comment by nickvergessen** [comment]
Scenarios I tested:

```
🗯 Message 1-150
🗯 Message 151  === lastReadMessage (from API)
🗯 Message 152-170
```
```
🗯 Message 1-150
🗯 Message 151  === lastReadMessage (from API)
🗯 Message 152-300
```
```
🗯 Message 1-150
😄 Reactions 1-150 on Message 150
🗯 Message 151  === lastReadMessage (from API)
🗯 Message 152-300
```
```
🗯 Message 1-150
🗯 Message 151  === lastReadMessage (from API)
😄 Reactions 1-150 on Message 151
🗯 Message 152-300
```
```
🗯 Message 1-150
😄 Reactions 1-150 on Message 150
😄 Reactions 150  === lastReadMessage (from API)
🗯 Message 152-300
```

All of them produce an acceptable state after loading the chat (better or equal to the previous behaviour). Some could still be improved in terms of scrolling off-set but that was already the case before, so I think we can keep it like that for now.

Any (edge) cases I forgot?

**Comment by danxuliu** [comment]
> Any (edge) cases I forgot?

I would also test the reaction scenarios with a deleted message instead of a reaction, and with a command from another participant that is not visible for the user loading the chat instead of a reaction.

**Comment by nickvergessen** [comment]
> I would also test the reaction scenarios with a deleted message instead of a reaction

Deleted messages are in the DOM, so they are not any different from normal messages

> and with a command from another participant that is not visible for the user loading the chat instead of a reaction.

Similar to messages that are not in the DOM but in the store, completely non-existing messages will just fall back to the previous visible one, so should be covered as well.

**Review by juliushaertl — APPROVED** [review]

**Comment by SystemKeeper** [comment]
Could this be used (or easily extended) to solve https://github.com/nextcloud/spreed/issues/6046 and be used in mobile clients?

**Comment by nickvergessen** [comment]
We are doing it in the web now, if you did not open the conversation before :)
But otherwise we can't as it might create a hole in the list of loaded messages :(

**Comment by SystemKeeper** [comment]
> But otherwise we can't as it might create a hole in the list of loaded messages :(

Could we remove the stored messages for that conversation when we open it from search to get the correct context ?

**Comment by nickvergessen** [comment]
Put as an idea into https://github.com/nextcloud/spreed/issues/6046 ?

**Pull request merged**